src/csim/stat_ops_expectation_value.cpp

```diff
@@ -260,15 +265,125 @@ double expectation_value_multi_qubit_Pauli_operator_partial_list_single_thread(
         Pauli_operator_type_list, target_qubit_index_count, &bit_flip_mask,
         &phase_flip_mask, &global_phase_90rot_count, &pivot_qubit_index);
     double result;
+    MPIutil& m = MPIutil::get_inst();
+
+#ifdef _OPENMP
+    OMPutil::get_inst().set_qulacs_num_threads(dim, 15);
+#endif
     if (bit_flip_mask == 0) {
-        result =
-            expectation_value_multi_qubit_Pauli_operator_Z_mask_single_thread(
-                phase_flip_mask, state, dim);
+        result = expectation_value_multi_qubit_Pauli_operator_Z_mask_mpi(
+            phase_flip_mask, state, dim, m.get_rank(), inner_qc);
     } else {
-        result =
-            expectation_value_multi_qubit_Pauli_operator_XZ_mask_single_thread(
-                bit_flip_mask, phase_flip_mask, global_phase_90rot_count,
-                pivot_qubit_index, state, dim);
+        result = expectation_value_multi_qubit_Pauli_operator_XZ_mask_mpi(
+            bit_flip_mask, phase_flip_mask, global_phase_90rot_count,
+            pivot_qubit_index, state, dim, inner_qc);
     }
+#ifdef _OPENMP
+    OMPutil::get_inst().reset_qulacs_num_threads();
+#endif
+
+    if (outer_qc > 0) MPIutil::get_inst().s_D_allreduce(&result);
     return result;
 }
+
+double expectation_value_multi_qubit_Pauli_operator_XZ_mask_mpi(
+    ITYPE bit_flip_mask, ITYPE phase_flip_mask, UINT global_phase_90rot_count,
+    UINT pivot_qubit_index, const CTYPE* state, ITYPE dim, UINT inner_qc) {
+    const ITYPE pivot_mask = 1ULL << pivot_qubit_index;
+    ITYPE state_index;
+    double sum = 0.;
+
+    int comm_flag = bit_flip_mask >> inner_qc;
+
+    MPIutil& m = MPIutil::get_inst();
+    int mpirank = m.get_rank();
+    int pair_rank = mpirank ^ comm_flag;
+    ITYPE global_offset = mpirank << inner_qc;
+
+    if (comm_flag) {
+        ITYPE dim_work = dim;
+        ITYPE num_work = 0;
+        CTYPE* recvptr = m.get_workarea(&dim_work, &num_work);
+        ITYPE inner_mask = dim - 1;
+        ITYPE i, j;
+
+        state_index = 0;
+
+        for (i = 0; i < num_work; ++i) {
+            const CTYPE* sendptr = state + dim_work * i;
+            if (mpirank < pair_rank) {
+                // recv
+                m.m_DC_recv(recvptr, dim_work, pair_rank);
+
+#ifdef _OPENMP
+#pragma omp parallel for reduction(+ : sum)
+#endif
+                for (j = 0; j < dim_work; ++j) {
+                    ITYPE basis_1 = state_index + j + (pair_rank << inner_qc);
+                    ITYPE basis_0 = basis_1 ^ bit_flip_mask;
+                    UINT sign_0 =
+                        count_population(basis_0 & phase_flip_mask) % 2;
+
+                    sum += _creal(
+                        state[basis_0 & inner_mask] *
+                        conj(recvptr[basis_1 & (dim_work - 1)]) *
```

Comment:
I think it is not guaranteed that basis_0 and basis_1 are in same `i`(workspace-index).

ex)
If
outer_qc=1
inner_qc=3
num_work=1<<1
bit_flip_mask=0b1101
,

an opponent of index 0b0000 (mpi_rank=0b0, i=0b0, j=0b00) is index 0b1101 (mpi_rank=0b1,i=0b1,j=0b01).

Reply:
This double loop of ```i``` and ```j``` is based on the sending side: the loop of ```i``` sweeps the communicating location in dim_work units.
Then, the loop of ```j```, which operates on the receiving side, scans the received chunks in order on the basis_1 side and obtains the corresponding basis_0 by exclusive-OR with the bit_flip_mask.

In other words, the sender accesses the state vector in order, while the receiver moves back and forth across its own state vector, ignoring the value of ```i```.

Reply:
Sorry. I was wrong. Thank you for explaining.